erts/emulator/beam/erl_process.c

```diff
@@ -7194,16 +7240,19 @@ resume_process(Process *p, ErtsProcLocks locks)
 	return;
 
     state = erts_atomic32_read_nob(&p->state);
+    if (state & ERTS_PSFLG_PAUSED_TIMER) {
+        schedule_unpause_proc_timer(p->common.id, locks);
```

Comment:
@rickard-green The i[nitial version of this PR](https://github.com/erlang/otp/commits/36bef84b16691500a3bea10e4a3ceaa7e19453ea) was using `erts_proc_sig_send_rpc_request()`. The reason I changed it to use sys-tasks was that at this point, we'd get a lock-order violation like this:

```
Lock order violation occurred when locking 'proc_msgq:<0.269.0>[proclock]'(none)!                                                                                                                                                                                                                                           
Currently these locks are locked by the scheduler 1 thread:                                                                                                                                                                                                                                                                 
  'proc_main:<0.269.0>[proclock](beam/erl_process.c:10109)'(none)                                                                                                                                                                                                                                                           
  'proc_status:<0.269.0>[proclock](beam/erl_process.c:9300)'(none) 
```

So the problem remains on how to avoid this. We need to somehow send the rpc-request signal before the status lock is acquired

Reply:
I missed that this one would have a problem as well.

The problem is that the unpausing is triggered from the low level primitive `resume_process()` which handles all resumes regardless of whether corresponding suspend came with a pause or not. Unpausing due to a resume that do not correspond to a suspend with a pause feels like a bug to me.

The only way to pause timers is via the `erlang:suspend_process()` BIF. If the unpausing is done from corresponding resume in the `erlang:resume_process()` BIF, the problem goes away.